cmd/mt-gateway/ingest/metrics.go

```diff
@@ -116,109 +117,124 @@ func prepareIngest(ctx *models.Context, in []*schema.MetricData, toPublish []*sc
 	return toPublish, resp
 }
 
-func metricsJson(ctx *models.Context) {
-	if ctx.Req.Request.Body == nil {
-		ctx.JSON(400, "no data included in request.")
+func metricsJson(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		w.WriteHeader(400)
+		fmt.Fprintf(w, "no data included in request.")
 		return
 	}
-	defer ctx.Req.Request.Body.Close()
-	body, err := ioutil.ReadAll(ctx.Req.Request.Body)
+	defer r.Body.Close()
+	body, err := ioutil.ReadAll(r.Body)
 	if err != nil {
 		select {
-		case <-ctx.Req.Context().Done():
-			ctx.Error(499, "request canceled")
+		case <-r.Context().Done():
+			w.WriteHeader(499)
+			fmt.Fprintf(w, "request canceled")
 		default:
 			log.Errorf("unable to read request body. %s", err)
-			ctx.JSON(500, err)
+			w.WriteHeader(500)
+			fmt.Fprintf(w,"unable to read request body. %s", err)
 		}
 		return
 	}
 	metrics := make([]*schema.MetricData, 0)
 	err = json.Unmarshal(body, &metrics)
 	if err != nil {
-		ctx.JSON(400, fmt.Sprintf("unable to parse request body. %s", err))
+		w.WriteHeader(400)
+		fmt.Fprintf(w, "unable to parse request body. %s", err)
 		return
 	}
 
 	toPublish := make([]*schema.MetricData, 0, len(metrics))
 	toPublish, resp := prepareIngest(ctx, metrics, toPublish)
 
 	select {
-	case <-ctx.Req.Context().Done():
-		ctx.Error(499, "request canceled")
+	case <-r.Context().Done():
+		w.WriteHeader(499)
+		fmt.Fprintf(w, "request canceled")
 		return
 	default:
 	}
 
 	err = publish.Publish(toPublish)
 	if err != nil {
 		log.Errorf("failed to publish metrics. %s", err)
-		ctx.JSON(500, err)
+		w.WriteHeader(500)
+		fmt.Fprintf(w, "failed to publish metrics. %s", err)
 		return
 	}
 
 	// track published in the response (it already has discards)
 	resp.Published = len(toPublish)
-	ctx.JSON(200, resp)
+	w.WriteHeader(200)
+	json.NewEncoder(w).Encode(resp)
 }
 
-func metricsBinary(ctx *models.Context, compressed bool) {
-	if ctx.Req.Request.Body == nil {
-		ctx.JSON(400, "no data included in request.")
+func metricsBinary(w http.ResponseWriter, r *http.Request, compressed bool) {
+	if r.Body == nil {
+		w.WriteHeader(400)
+		fmt.Fprintf(w, "no data included in request.")
 		return
 	}
 	var bodyReadCloser io.ReadCloser
 	if compressed {
-		bodyReadCloser = ioutil.NopCloser(snappy.NewReader(ctx.Req.Request.Body))
+		bodyReadCloser = ioutil.NopCloser(snappy.NewReader(r.Body))
 	} else {
-		bodyReadCloser = ctx.Req.Request.Body
+		bodyReadCloser = r.Body
 	}
 	defer bodyReadCloser.Close()
 
 	body, err := ioutil.ReadAll(bodyReadCloser)
 	if err != nil {
 		select {
-		case <-ctx.Req.Context().Done():
-			ctx.Error(499, "request canceled")
+		case <-r.Context().Done():
+			w.WriteHeader(499)
+			fmt.Fprintf(w, "request canceled")
 		default:
 			log.Errorf("unable to read request body. %s", err)
-			ctx.JSON(500, err)
+			w.WriteHeader(500)
+			fmt.Fprintf(w, "unable to read request body. %s", err)
```

Comment:
there's a repeating pattern of writing a status code header, then writing a message on the following line, and in some cases also log a line with the same msg. 
would it make sense to put these into a small helper like `writeStatus(w, 499, "request cancelled", true)` (the last param could determine whether to log).

Reply:
Yeah, that had crossed my mind.
Skipped the  boolean and just always log @ error (successful ingestions take a different code path)

1abc592486eb7c50c4d2b25b84f73d6c037704dc